posix: sched: ensure min and max priority are schedulable

cfriedt opened the pull request:
Previously, with preemptive scheduling, minimum priority threads were not schedulable. The linked issue appears to highlight an off-by-one error as well, which I believe should be fixed here.

Also add `SCHED_OTHER` support, which is mandatory (separate commits)

Fixes #56729

Commit:
posix: sched: ensure min and max priority are schedulable

Previously, there was an off-by-one error for SCHED_RR.

Fixes #56729

Signed-off-by: Chris Friedt <cfriedt@meta.com>

Commit:
tests: posix: ensure that min and max priority are schedulable

Verify that threads are actually schedulable for min and max
scheduler priority for both `SCHED_RR` (preemptive) and
`SCHED_FIFO` (cooperative).

Fixes #56729

Signed-off-by: Chris Friedt <cfriedt@meta.com>

cfriedt commented:
Cc @ambroise-arm

jgl-meta commented on lib/posix/pthread_sched.c:
```diff
@@ -51,25 +34,10 @@ int sched_get_priority_min(int policy)
+	if (IS_ENABLED(CONFIG_COOP_ENABLED) && policy == SCHED_FIFO) {
+		return CONFIG_NUM_COOP_PRIORITIES - 1;
+	} else if (IS_ENABLED(CONFIG_PREEMPT_ENABLED) && policy == SCHED_RR) {
+		return CONFIG_NUM_PREEMPT_PRIORITIES - 1;
```
Why does the preempt priorities now need a -1?

cfriedt commented on lib/posix/pthread.c:
```diff
@@ -89,7 +87,7 @@ static int32_t posix_to_zephyr_priority(uint32_t priority, int policy)
 		prio =  -1 * (priority + 1);
 	} else {
-		prio = (CONFIG_NUM_PREEMPT_PRIORITIES - priority);
+		prio = (CONFIG_NUM_PREEMPT_PRIORITIES - 1 - priority);
```
I'm fine doing another rev. Might make sense to use a loop too, just to be less verbose

cfriedt commented on tests/posix/common/src/pthread.c:
```diff
@@ -598,3 +606,128 @@ ZTEST(posix_apis, test_pthread_descriptor_leak)
+		 *
+		 * Note: we relax this requirement
+		 */
+		zassert_true(pmax > pmin, "pmax (%d) <= pmin (%d)", pmax, pmin);
```
It's actually not possible to support `pmax >= pmin` because that would mean `CONFIG_<policy>_ENABLED=n`.

I did switch to a for-loop though.

Commit:
posix: sched: add support for SCHED_OTHER

The `SCHED_OTHER` scheduling priority is mandatory as part of
POSIX. It must be numerically distinct from `SCHED_FIFO`,
`SCHED_RR`, and `SCHED_SPORADIC`, but is implementation-
defined and may behave identically to `SCHED_FIFO` or
`SCHED_RR`.

Signed-off-by: Chris Friedt <cfriedt@meta.com>

Commit:
tests: posix: test support for SCHED_OTHER

Ensure we test `SCHED_OTHER` functionality along with other
scheduling policies and priorities.

Signed-off-by: Chris Friedt <cfriedt@meta.com>

cfriedt commented on lib/posix/pthread.c:
```diff
@@ -14,16 +14,15 @@
 #include <zephyr/sys/slist.h>
 
 #include "posix_internal.h"
```
Just a note, but the `posix_internal.h` header is unfortunately sensitive to include order 😢 - this should be fixed at some point (#57232)

cfriedt commented:
Oh wow - something broke all of the tests.. it's not even rendering in GH 😅

stephanosio commented:
> Oh wow - something broke all of the tests.. it's not even rendering in GH sweat_smile

```
In file included from /__w/zephyr/zephyr/include/zephyr/posix/pthread.h:15,
                 from /__w/zephyr/zephyr/tests/posix/common/src/mqueue.c:13:
/__w/zephyr/zephyr/include/zephyr/posix/sched.h:26: error: "SCHED_OTHER" redefined [-Werror]
   26 | #define SCHED_OTHER 3
      | 
In file included from /opt/toolchains/zephyr-sdk-0.16.0/arm-zephyr-eabi/arm-zephyr-eabi/sys-include/sys/_pthreadtypes.h:23,
                 from /opt/toolchains/zephyr-sdk-0.16.0/arm-zephyr-eabi/arm-zephyr-eabi/sys-include/sys/types.h:223,
                 from /opt/toolchains/zephyr-sdk-0.16.0/arm-zephyr-eabi/arm-zephyr-eabi/sys-include/stdio.h:61,
                 from /__w/zephyr/zephyr/subsys/testsuite/ztest/include/zephyr/ztest_assert.h:18,
                 from /__w/zephyr/zephyr/subsys/testsuite/ztest/include/zephyr/ztest.h:49,
                 from /__w/zephyr/zephyr/tests/posix/common/src/mqueue.c:7:
/opt/toolchains/zephyr-sdk-0.16.0/arm-zephyr-eabi/arm-zephyr-eabi/sys-include/sys/sched.h:35: note: this is the location of the previous definition
   35 | #define SCHED_OTHER    0
      | 
cc1: all warnings being treated as errors
```

cfriedt commented:
> /opt/toolchains/zephyr-sdk-0.16.0/arm-zephyr-eabi/arm-zephyr-eabi/sys-include/sys/sched.h:35: note: this is the location of the previous definition
>    35 | #define SCHED_OTHER    0

You win this time ARM toolchain...

cfriedt commented:
@jgl-meta, @stephanosio, @keith-packard - should be good to go if you're up for reviewing

cfriedt commented:
* reuse `create_thread1()` instead of defining a second no-op pthread function

stephanosio reviewed (approved):
I am unable to find anything obviously wrong with the changes in this PR.
